### ISSUE TITLE
PropOptions causing Typescript-Error (TS2345)

### ISSUE THREAD

**Issue by SvenTiigi** [nestjs/mongoose#752]
## I'm submitting a...

<pre><code>
[ ] Regression 
[x] Bug report
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior
Currently with the latest release of `@nestjs/mongoose` (7.2.2) Typescript no longer compiles any schema object that has properties with the `@Prop` decorator with a `PropOptions` object.

For example this code snippet (taken from the official NestJS documentation)
[https://docs.nestjs.com/techniques/mongodb](https://docs.nestjs.com/techniques/mongodb)

```ts
import {Prop, Schema} from "@nestjs/mongoose";

@Schema()
export class Cat {

    @Prop({ required: true })
    name: string;

}
```

will generate the following Typescript transpiling error:

```bash
error TS2345: Argument of type '{ required: true; }' is not assignable to parameter of type 'PropOptions'.
  Type '{ required: true; }' is not assignable to type '"number"'.

19     @Prop({required: true})
             ~~~~~~~~~~~~~~~~
```

#### Package.json

<details>
  <pre>
  "dependencies": {
    "@nestjs/common": "^7.5.1",
    "@nestjs/config": "^0.6.3",
    "@nestjs/core": "^7.5.1",
    "@nestjs/jwt": "^7.2.0",
    "@nestjs/mongoose": "^7.2.0",
    "@nestjs/passport": "^7.1.5",
    "@nestjs/platform-express": "^7.5.1",
    "@nestjs/swagger": "^4.7.12",
    "bcrypt": "^5.0.0",
    "class-transformer": "^0.3.2",
    "class-validator": "^0.13.1",
    "compression": "^1.7.4",
    "helmet": "^4.4.1",
    "mongoose": "^5.11.15",
    "nestjs-redoc": "^2.1.1",
    "passport-jwt": "^4.0.0",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^6.6.3",
    "swagger-ui-express": "^4.1.6"
  },
  "devDependencies": {
    "@nestjs/cli": "^7.5.1",
    "@nestjs/schematics": "^7.1.3",
    "@nestjs/testing": "^7.5.1",
    "@types/bcrypt": "^3.0.0",
    "@types/compression": "^1.7.0",
    "@types/express": "^4.17.8",
    "@types/helmet": "^4.0.0",
    "@types/jest": "^26.0.15",
    "@types/node": "^14.14.6",
    "@types/passport-jwt": "^3.0.4",
    "@types/supertest": "^2.0.10",
    "@typescript-eslint/eslint-plugin": "^4.6.1",
    "@typescript-eslint/parser": "^4.6.1",
    "eslint": "^7.12.1",
    "eslint-config-prettier": "7.0.0",
    "eslint-plugin-prettier": "^3.1.4",
    "jest": "^26.6.3",
    "prettier": "^2.1.2",
    "supertest": "^6.0.0",
    "ts-jest": "^26.4.3",
    "ts-loader": "^8.0.8",
    "ts-node": "^9.0.0",
    "tsconfig-paths": "^3.9.0",
    "typescript": "^4.0.5"
  },
</pre>
</details>

I have an older NestJS project which is still compiling and uses the `@nestjs/mongoose` package in version `7.0.2`. In this project the `PropOptions` type is declared in the following way (which causes no errors):

```ts
export declare type PropOptions = mongoose.SchemaTypeOpts<any> | mongoose.Schema | mongoose.SchemaType;
```

In the current version `7.2.2` of `@nestjs/mongoose` the `PropOptions` is declared in this way (which is causing the Typescript-Error):

```ts
export declare type PropOptions = mongoose.SchemaDefinition['string'] | mongoose.SchemaType;
```

## Expected behavior
No Typescript transpiling error occurs.

## Minimal reproduction of the problem with instructions

[https://github.com/SvenTiigi/nestjs-mongoose-issue-752.git](https://github.com/SvenTiigi/nestjs-mongoose-issue-752.git)

or:

Generate a new NestJS Project and add the `@nestjs/mongoose` package:

```bash
$ nest new test-project
$ cd test-project
$ npm i --save @nestjs/mongoose mongoose
```

add the following code snippet to `app.module.ts`

```ts
import {Prop, Schema} from "@nestjs/mongoose";

@Schema()
export class Cat {

    @Prop({ required: true })
    name: string;

}
```

Run `npm start`

## Environment

<pre><code>
Nest version: 7.5.4

 
For Tooling issues:
- Node version: v15.7.0
- Platform: Mac

Others:

</code></pre>

**Comment by woodstock13** [comment]
Hi, same here I just rollback to mongoose: 5.11.14 to quick fix.

**Comment by Tony133** [comment]
Hi @SvenTiigi , I advise you to leave a minimal reproduction of a clonable git repository so that the core team can evaluate the problem you have reported.

**Comment by SvenTiigi** [comment]
Hi @Tony133

Here is a minimal clonable git repository used to reproduce the described issue:

[https://github.com/SvenTiigi/nestjs-mongoose-issue-752](https://github.com/SvenTiigi/nestjs-mongoose-issue-752)

**Comment by jmcdo29** [comment]
Looks like [there was an update to the `index.d.ts` for mongoose that caused this](https://github.com/Automattic/mongoose/commit/bdbf25d137ad3a715adf973e919be4897eeec445)

**Comment by Tony133** [comment]
It could be fixed like this but I don't know if it is correct

`export declare type PropOptions = mongoose.SchemaDefinition['string'] | any | mongoose.SchemaType;`

**Comment by jmcdo29** [comment]
If we can avoid `any` we should. If we need it, we should _probably_ use `unknown` instead

**Comment by Tony133** [comment]
with `any` or with `unknown` the given example works, and gives no error

**Comment by kamilmysliwiec** [comment]
Fixed in v7.2.3
